File: Python/7-Trees/Problems/1-InvertBinaryTree/InvertBinaryTreePro.py
from collections import deque
from typing import Optional, Dict
import time
import pickle
import heapq

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val  # Store the value of the node.
        self.left = left  # Store the left child node.
        self.right = right  # Store the right child node.

# Approach 1: Time Delay Simulation (Simulating Processing Time)
def invert_tree_with_delay(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Inverts a binary tree by swapping the left and right children of each node.
    This version simulates processing time with a delay.

    Args:
        root: The root of the binary tree to invert.

    Returns:
        The root of the inverted binary tree.
    """
    if not root:  # Base case: if the node is None, return None.
        return None
    time.sleep(0.1)  # Simulate processing time (100ms delay).  This is NOT a standard part of tree inversion.
    root.left, root.right = root.right, root.left  # Swap left and right children.
    invert_tree_with_delay(root.left)  # Recursively invert the left subtree.
    invert_tree_with_delay(root.right)  # Recursively invert the right subtree.
    return root  # Return the root of the modified tree.

# Approach 2: Binary Heap Storage and Inversion
def invert_tree_heap(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Inverts a binary tree using a binary heap.  This is an unusual and inefficient approach.
    It stores node values in a heap, then reconstructs the tree from the heap.  This does NOT
    preserve the original tree structure, and the inverted tree will be heap-ordered, not
    a true inversion.

    Args:
        root: The root of the binary tree.

    Returns:
        The root of the "inverted" tree (which will be heap-ordered).
    """
    if not root:
        return None

    heap = []  # Initialize an empty list to be used as a min-heap.

    def store_in_heap(node: Optional[TreeNode]):
        """
        Recursively traverses the tree and stores the *values* of the nodes in the heap.
        """
        if node:
            heapq.heappush(heap, node.val)  # Push the node's value onto the heap.
            store_in_heap(node.left)  # Recurse on the left subtree.
            store_in_heap(node.right)  # Recurse on the right subtree.

    store_in_heap(root)  # Start storing node values from the root.

    def construct_inverted_tree():
        """
        Recursively constructs a new tree from the values in the heap.  The tree will be
        heap-ordered (smallest value at the root), NOT a proper inversion.  The original
        tree structure is lost.
        """
        if not heap:
            return None
        return TreeNode(heapq.heappop(heap), construct_inverted_tree(), construct_inverted_tree())  # Create a new node with the smallest value from the heap.

    return construct_inverted_tree()  # Start constructing the new tree.

# Approach 3: Pickle Serialization and Inversion
def invert_tree_pickle(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Inverts a binary tree using Python's pickle serialization.  This is also an unusual
    and generally inefficient way to invert a binary tree.  It serializes the tree,
    deserializes it, and then swaps the left and right children at the root of the *deserialized* tree.
    This will not correctly invert subtrees.

    Args:
        root: The root of the binary tree.

    Returns:
        The root of the inverted tree.  Inversion is only done at the top level.
    """
    if not root:
        return None
    serialized_tree = pickle.dumps(root)  # Serialize the tree into a byte string.
    inverted_tree = pickle.loads(serialized_tree)  # Deserialize the byte string back into a tree structure.
    inverted_tree.left, inverted_tree.right = inverted_tree.right, inverted_tree.left # Only swaps the root's children.
    return inverted_tree

# Approach 4: Stack-based Manual Inversion
def invert_tree_stack(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Inverts a binary tree using a stack (iterative approach).  This is a standard and
    efficient way to invert a binary tree.

    Args:
        root: The root of the binary tree.

    Returns:
        The root of the inverted tree.
    """
    if not root:
        return None
    stack = [root]  # Initialize a stack with the root node.
    while stack:
        node = stack.pop()  # Pop a node from the stack.
        node.left, node.right = node.right, node.left  # Swap left and right children.
        if node.left:
            stack.append(node.left)  # Push the left child onto the stack.
        if node.right:
            stack.append(node.right)  # Push the right child onto the stack.
    return root  # Return the root of the modified tree.

# Approach 5: Inversion with Linked List Simulation (using deque)
def invert_tree_linked_list(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Inverts a binary tree using a deque (double-ended queue) to simulate a queue
    (breadth-first traversal). This is a standard and efficient way to invert a binary tree.
    This is essentially an iterative level-order traversal where you swap children.

    Args:
        root: The root of the binary tree.

    Returns:
        The root of the inverted tree.
    """
    if not root:
        return None
    queue = deque([root])  # Initialize a deque with the root node.
    while queue:
        node = queue.popleft()  # Dequeue a node from the front.
        node.left, node.right = node.right, node.left  # Swap left and right children.
        if node.left:
            queue.append(node.left)  # Enqueue the left child.
        if node.right:
            queue.append(node.right)  # Enqueue the right child.
    return root  # Return the root of the modified tree.

# Approach 6: Counting Nodes and Inverting
def invert_tree_counting(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Inverts a binary tree and counts the number of nodes before and after inversion.
    This is not a different inversion method, but adds a node-counting operation
    to the standard swapping of left and right children.

    Args:
        root: The root of the binary tree.

    Returns:
        The root of the inverted tree.
    """
    if not root:
        return None

    def count_nodes(node: Optional[TreeNode]) -> int:
        """
        Recursively counts the number of nodes in the subtree rooted at 'node'.
        """
        if not node:
            return 0
        return 1 + count_nodes(node.left) + count_nodes(node.right)  # Count the current node + left subtree + right subtree.

    print(f"Total nodes before inversion: {count_nodes(root)}")  # Print the node count before inversion.
    root.left, root.right = root.right, root.left  # Swap left and right children.
    print(f"Total nodes after inversion: {count_nodes(root)}")  # Print the node count after inversion.
    return root  # Return the root of the inverted tree.

# Helper function to print tree (Breadth-First Traversal)
def print_tree(root: Optional[TreeNode]):
    """
    Prints the binary tree in a breadth-first manner (level-order traversal).
    This function prints node values, and None for missing nodes, in a list format.

    Args:
        root: The root of the binary tree to print.
    """
    if not root:
        return
    queue = deque([root])  # Initialize a deque with the root node.
    result = []  # List to store the node values.
    while queue:
        node = queue.popleft()  # Dequeue a node.
        result.append(node.val if node else None)  # Append node's value or None if node is None
        if node:
            queue.append(node.left)  # Enqueue left child.
            queue.append(node.right)  # Enqueue right child.
    print(result)  # Print the list of node values.

# Sample Usage
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
print("Original Tree:")
print_tree(tree)

print("\nInverted using Time Delay:")
print_tree(invert_tree_with_delay(tree))

# Recreate the tree
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
print("\nInverted using Binary Heap:")
print_tree(invert_tree_heap(tree))

# Recreate the tree
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
print("\nInverted using Pickle Serialization:")
print_tree(invert_tree_pickle(tree))

# Recreate the tree
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
print("\nInverted using Stack:")
print_tree(invert_tree_stack(tree))

# Recreate the tree
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
print("\nInverted using Linked List Simulation:")
print_tree(invert_tree_linked_list(tree))

# Recreate the tree
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
print("\nInverted using Node Counting:")
print_tree(invert_tree_counting(tree))

File: C++/7-Trees/Problems/1-InvertBinaryTree/InvertBinaryTreePro.js
/**
 * Invert Binary Tree - 5 Approaches
 *
 * A binary tree node has a value, a left child, and a right child.
 *
 * 1. Recursive Approach
 * 2. Iterative Approach (using a stack)
 * 3. Iterative Approach (using a queue)
 * 4. Recursive Approach (Simplified)
 * 5. Morris Traversal (Threaded Tree) - [Advanced, for discussion]
 */

// Definition of a binary tree node.
function TreeNode(val, left, right) {
  this.val = (val === undefined ? 0 : val);
  this.left = (left === undefined ? null : left);
  this.right = (right === undefined ? null : right);
}

// Helper function to create a sample binary tree for testing
function createSampleTree() {
  const root = new TreeNode(4);
  root.left = new TreeNode(2);
  root.right = new TreeNode(7);
  root.left.left = new TreeNode(1);
  root.left.right = new TreeNode(3);
  root.right.left = new TreeNode(6);
  root.right.right = new TreeNode(9);
  return root;
}

// Helper function to perform an inorder traversal (Left, Root, Right)
// Used to verify the tree structure after inversion
function inorderTraversal(root) {
  const result = [];
  if (root) {
    result.push(...inorderTraversal(root.left));
    result.push(root.val);
    result.push(...inorderTraversal(root.right));
  }
  return result;
}

// Helper function to visualize the tree structure (for debugging)
function visualizeTree(root, indent = "") {
  if (root) {
    console.log(indent + root.val);
    visualizeTree(root.left, indent + "  ");
    visualizeTree(root.right, indent + "  ");
  }
}

/**
 * 1. Recursive Approach
 *
 * - Base Case: If the node is null, return.
 * - Recursive Step:
 * - Swap the left and right children of the current node.
 * - Recursively call the function for the left and right children.
 *
 * Time Complexity: O(n), where n is the number of nodes in the tree.
 * Space Complexity: O(h), where h is the height of the tree (due to the call stack).  In the worst case (skewed tree), h = n, so O(n).  In the best case (balanced tree), h = log n, so O(log n).
 */
function invertTreeRecursive(root) {
  if (!root) {
    return null;
  }

  // Swap left and right children
  const temp = root.left;
  root.left = root.right;
  root.right = temp;

  // Recursively invert the left and right subtrees
  invertTreeRecursive(root.left);
  invertTreeRecursive(root.right);

  return root;
}

/**
 * 2. Iterative Approach (using a stack)
 *
 * - Use a stack to store nodes to be processed.
 * - Push the root node onto the stack.
 * - While the stack is not empty:
 * - Pop a node from the stack.
 * - Swap the left and right children of the node.
 * - Push the left and right children (if they exist) onto the stack.
 *
 * Time Complexity: O(n), where n is the number of nodes in the tree.
 * Space Complexity: O(h), where h is the height of the tree.  In the worst case (skewed tree), h = n, so O(n). In the best case (balanced tree), h = log n, so O(log n).
 */
function invertTreeIterativeStack(root) {
  if (!root) {
    return null;
  }

  const stack = [root];

  while (stack.length > 0) {
    const node = stack.pop();

    // Swap left and right children
    const temp = node.left;
    node.left = node.right;
    node.right = temp;

    // Push children onto the stack
    if (node.left) {
      stack.push(node.left);
    }
    if (node.right) {
      stack.push(node.right);
    }
  }

  return root;
}

/**
 * 3. Iterative Approach (using a queue)
 *
 * - Use a queue to store nodes to be processed.
 * - Enqueue the root node.
 * - While the queue is not empty:
 * - Dequeue a node from the queue.
 * - Swap the left and right children of the node.
 * - Enqueue the left and right children (if they exist) into the queue.
 *
 * Time Complexity: O(n), where n is the number of nodes in the tree.
 * Space Complexity: O(w), where w is the maximum width of the tree. In the worst case (complete binary tree), w = n/2, so O(n). In the best case (skewed tree), w = 1, so O(1).
 */
function invertTreeIterativeQueue(root) {
  if (!root) {
    return null;
  }

  const queue = [root];

  while (queue.length > 0) {
    const node = queue.shift(); // Dequeue

    // Swap left and right children
    const temp = node.left;
    node.left = node.right;
    node.right = temp;

    // Enqueue children
    if (node.left) {
      queue.push(node.left);
    }
    if (node.right) {
      queue.push(node.right);
    }
  }

  return root;
}

/**
 * 4. Recursive Approach (Simplified)
 *
 * This is a more concise version of the recursive approach.
 */
function invertTreeRecursiveSimplified(root) {
  if (!root) return null;
  [root.left, root.right] = [invertTreeRecursiveSimplified(root.right), invertTreeRecursiveSimplified(root.left)];
  return root;
}

/**
 * 5. Morris Traversal (Threaded Tree) - [Advanced, for discussion]
 *
 * - This approach inverts the tree in-place without using recursion or a stack/queue.
 * - It modifies the tree structure temporarily by creating "threads" (links from a node's right child to its inorder successor).
 * - After inversion, the threads are removed to restore the original tree structure.
 * - This method is more space-efficient (O(1) space) but is more complex to implement.
 *
 * Time Complexity: O(n), where n is the number of nodes in the tree.
 * Space Complexity: O(1) - Constant space.
 *
 * **Note:** Morris traversal modifies the tree structure during the process.  It's generally not recommended for scenarios where you need to preserve the original tree structure, unless you make a copy first.  This is more of an academic exercise or an optimization for very specific use cases where memory is extremely limited and in-place modification is acceptable.  It's more complex and harder to debug.
 */
function invertTreeMorris(root) {
  let current = root;
  while (current) {
    if (current.left) {
      let predecessor = current.left;
      while (predecessor.right && predecessor.right !== current) {
        predecessor = predecessor.right;
      }

      if (!predecessor.right) {
        // Create the thread
        predecessor.right = current;

        // Move to the left child
        current = current.left;
      } else {
        // Remove the thread
        predecessor.right = null;

        // Swap children and move to the right child
        const temp = current.left;
        current.left = current.right;
        current.right = temp;
        current = current.right;
      }
    } else {
      // No left child, swap children and move to the right
      const temp = current.left;
      current.left = current.right;
      current.right = temp;
      current = current.right;
    }
  }
  return root;
}



// --- Test Cases and Output ---
const initialTree = createSampleTree();

console.log("Original Tree (Visualization):");
visualizeTree(initialTree);
console.log("Original Tree (Inorder Traversal):", inorderTraversal(initialTree));
console.log("\n");

// 1. Recursive Approach
const tree1 = createSampleTree(); // Get a fresh copy
const invertedTree1 = invertTreeRecursive(tree1);
console.log("1. Recursive Inversion (Visualization):");
visualizeTree(invertedTree1);
console.log("1. Recursive Inversion (Inorder Traversal):", inorderTraversal(invertedTree1));
console.log("\n");

// 2. Iterative Approach (Stack)
const tree2 = createSampleTree();  // Get a fresh copy
const invertedTree2 = invertTreeIterativeStack(tree2);
console.log("2. Iterative Inversion (Stack) (Visualization):");
visualizeTree(invertedTree2);
console.log("2. Iterative Inversion (Stack) (Inorder Traversal):", inorderTraversal(invertedTree2));
console.log("\n");

// 3. Iterative Approach (Queue)
const tree3 = createSampleTree(); // Get a fresh copy
const invertedTree3 = invertTreeIterativeQueue(tree3);
console.log("3. Iterative Inversion (Queue) (Visualization):");
visualizeTree(invertedTree3);
console.log("3. Iterative Inversion (Queue) (Inorder Traversal):", inorderTraversal(invertedTree3));
console.log("\n");

// 4. Recursive Approach (Simplified)
const tree4 = createSampleTree();
const invertedTree4 = invertTreeRecursiveSimplified(tree4);
console.log("4. Recursive Inversion (Simplified) (Visualization):");
visualizeTree(invertedTree4);
console.log("4. Recursive Inversion (Simplified) (Inorder Traversal):", inorderTraversal(invertedTree4));
console.log("\n");

// 5. Morris Traversal (Threaded Tree)
const tree5 = createSampleTree();
const invertedTree5 = invertTreeMorris(tree5);
console.log("5. Morris Traversal Inversion (Visualization):");
visualizeTree(invertedTree5);
console.log("5. Morris Traversal Inversion (Inorder Traversal):", inorderTraversal(invertedTree5));
console.log("\n");
